packages/expo-clipboard/android/src/main/java/expo/modules/clipboard/ClipboardImage.kt

```diff
@@ -0,0 +1,171 @@
+ *
+ * Saves the image to the [clipboardCacheDir] directory, then uses [ClipboardFileProvider]
+ * to create a `content://` URI which then is packed to the [ClipData] object.
+ *
+ * @param context
+ * @param base64Image base64-encoded JPEG image data. Should not be prefixed
+ * @param clipboardCacheDir directory where the copied image is stored, must be accessible by
+ * the [ClipboardFileProvider]
+ * @return clip data ready to be shared by the [android.content.ClipboardManager]
+ */
+internal suspend fun clipDataFromBase64Image(
+  context: Context,
+  base64Image: String,
+  clipboardCacheDir: File,
+): ClipData {
+  // 1. Get bitmap from base64 string
+  val data = maybeStripBase64Prefix(base64Image)
+  val bitmap = bitmapFromBase64String(data)
+
+  // 2. Create file in cache dir, it will be overwritten if already exists
+  val file = File(clipboardCacheDir, "copied_image.jpeg").also {
+    it.ensureExists()
+  }
+
+  // 3. Write bitmap to the file
+  val fileStream = runInterruptible { FileOutputStream(file, false) }
+  BufferedOutputStream(fileStream).use { outputStream ->
+    bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream)
+    runInterruptible { outputStream.flush() }
+  }
+
+  // 4. Get content:// URI to the image file and put it to the clipboard data
+  val imageUri = ClipboardFileProvider.getUriForFile(
+    context,
+    context.applicationInfo.packageName + ".ClipboardFileProvider",
+    file
+  )
+  return ClipData.newUri(context.contentResolver, "image", imageUri)
+}
+// endregion
+
+// region Utility functions
+
+/**
+ * Retrieves [Bitmap] from `content://` image uri
+ * @throws SecurityException when app has no permission to access the content uri
+ * @throws IOException
+ */
+private suspend fun bitmapFromContentUriAsync(context: Context, imageUri: Uri): Bitmap =
+  runInterruptible(Dispatchers.IO) {
+    val contentResolver = context.contentResolver
+    when {
+      Build.VERSION.SDK_INT < 28 -> MediaStore.Images.Media.getBitmap(
+        contentResolver,
+        imageUri
+      )
+      else -> {
+        val source = ImageDecoder.createSource(contentResolver, imageUri)
+        ImageDecoder.decodeBitmap(source)
+      }
+    }
+  }
+
+private fun bitmapFromBase64String(base64Image: String): Bitmap = try {
+  val byteArray = Base64.decode(base64Image, Base64.DEFAULT)
+  BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
+    ?: throw RuntimeException("Failed to convert base64 into Bitmap")
+} catch (e: RuntimeException) {
+  // Base64.decode throws IllegalArgumentException on invalid data, but
+  // BitmapFactory.decodeByteArray returns null if data cannot be converted
+  // so we aggregate both errors here
+  throw InvalidImageException(base64Image, e)
+}
+
+/**
+ * Strips the `data:image/...;base64.` prefix from the base64 string
+ * leaving the pure base64 data string.
+ * Does nothing if there's no such prefix.
+ */
+private fun maybeStripBase64Prefix(base64String: String): String =
```

Comment:
we should do it on iOS too - or maybe move this to the JS side? We have documented that we accept strings _without_ the prefix, but checking it anyway makes it more foolproof

Reply:
> Ultimately our main focus should be on preventing runtime crash of the whole app. Would it crash for the malformed strings?

No, it just throws an exception or (if the malformed data is still a valid base64 or convertible automatically) returns corrupted image.

Reply:
Great, so I can just remove this as the `base64 -> Bitmap` already throws proper exception when input cannot be processed 🎉